clang/tools/driver/cc1depscand_main.cpp

```diff
@@ -797,14 +801,19 @@ static void updateCompilerInvocation(CompilerInvocation &Invocation,
                                      llvm::cas::CachingOnDiskFileSystem &FS,
                                      std::string RootID,
                                      StringRef CASWorkingDirectory,
+                                     Optional<CASOptions> CASOpts,
                                      llvm::TreePathPrefixMapper &Mapper) {
   // "Fix" the CAS options.
   auto &FileSystemOpts = Invocation.getFileSystemOpts();
   FileSystemOpts.CASFileSystemRootID = RootID;
   FileSystemOpts.CASFileSystemWorkingDirectory = CASWorkingDirectory.str();
   auto &FrontendOpts = Invocation.getFrontendOpts();
   FrontendOpts.CacheCompileJob = true;      // FIXME: Don't set.
-  Invocation.getCASOpts().CASPath = "auto"; // FIXME: Don't override.
+  // FIXME: Make passing CASOpts required.
+  if (CASOpts)
+    Invocation.getCASOpts() = *CASOpts;
```

Comment:
Should be `std::move()` I think.

Reply:
I think the logic we want is:
```
if (Invocation.getCASOpts().getKind() == CASOptions::InMemory)
  Invocation.getCASOpts().Path = "auto";
```

Reply:
Okay, after looking at the code in more detail (and what information the daemon has / should have), here's what I think we should do:

1. NFC refactor so that `updateCC1Args()` takes an `Optional<CASOptions>` that's used to override. Code is `if (OverrideCASOpts) Invocation.CASOpts = OverrideCASOpts;`. Updates both `-cc1depscan` and `-cc1depscand` to pass in one that has `-fcas-path=auto` (daemon uses the new CASOptions variable to construct its CAS).
2. `-cc1depscan` stops sending in `CASOptions`. Now you can add a test where the options are passed to `-cc1-args` and respected. No need to have `-cc1depscan` duplicate them on teh command-line since it's just getting them from `-cc1-args`.
3. Caller of the daemon starts sniffing `CASOptions` from the CC1Args, replacing with `-fcas-path=auto` when necessary (if it was in-memory). `makeDepscanDaemonKey()` is changed to take `CASOptions`, and serializes/hashes them so that different CASOptions mean a different daemon. The `-cc1depscand` command-line takes the `CASOptions` on the command-line instead of always using `-fcas-path=auto`.

End state is that the daemon overrides the CASOptions (since it needs a single consistent CAS for use with CachingOnDiskFileSystem), but non-daemon does not (just uses whatever was passed on `-cc1-args`).

Reply:
CASOptions are not on -cc1-args. They are before `-cc1-args`. 

I agree with the suggestion that we can hash CASOption to create unique daemon based on CASOption.

Reply:
I like the suggestions about how to change the daemon but I'd prefer that we enforce that the caller determines the CAS options and passes them to `updateCC1Args()` for dependency scanner to use (so the goal should be that it is non-optional). This ensures there is no chance for a mismatch of what CAS usage the caller expects and what dependency scanning is producing. `-cc1depscan` could pick up relevant options from `-cc1-args` and pass them.

Reply:
> CASOptions are not on -cc1-args. They are before -cc1-args.

`CASOptions` is definitely in `-cc1-args`. That's how `-cc1` gets them.

Fixing this mess is what's blocking adding `CASOptions` to the driver, BTW.



> (I'm going to do the NFC patch and post it in a bit)

Sorry I failed to actually post this comment when I wrote it over an hour ago :(. Anyway, the NFC commit is almost done, just making sure tests pass.

> I like the suggestions about how to change the daemon but I'd prefer that we enforce that the caller determines the CAS options and passes them to `updateCC1Args()` for dependency scanner to use (so the goal should be that it is non-optional).

You're right. In practice that's what made sense when writing the NFC refactoring. There is a failing test but I'll post in a moment.

Reply:
Refactoring in https://github.com/apple/llvm-project/pull/4127. I left behind FIXMEs for the other steps. I also suggested an order for the follow ups in that commit (start with daemon, then non-daemon, then (finally) add to driver).

Reply:
That landed, and I did a bunch more refactoring to clean things up further. #4130 is a PR for the last bit; you should rebase on top of that (I'll land once I notice that tests pass).